File: harness/c/checkpoint_harness.c
#include <stdint.h>

#include "harness_abi.h"
#include "../../c/checkpoint/checkpoint.h"

_Static_assert((int)CHECKPOINT_RESTART_COMMITTED == HARNESS_RESTART_COMMITTED,
    "checkpoint_restart_status_t::CHECKPOINT_RESTART_COMMITTED must match HARNESS_RESTART_COMMITTED");
_Static_assert((int)CHECKPOINT_RESTART_RESTORED == HARNESS_RESTART_RESTORED,
    "checkpoint_restart_status_t::CHECKPOINT_RESTART_RESTORED must match HARNESS_RESTART_RESTORED");
_Static_assert((int)CHECKPOINT_RESTART_RESTORE_FAILED == HARNESS_RESTART_RESTORE_FAILED,
    "checkpoint_restart_status_t::CHECKPOINT_RESTART_RESTORE_FAILED must match HARNESS_RESTART_RESTORE_FAILED");

volatile uint32_t harness_iteration;
volatile uint32_t harness_stage;
volatile uint32_t harness_fault_target;
volatile uint32_t harness_fault_value;
volatile uint32_t harness_last_expected;
volatile uint32_t harness_last_initial_value;
volatile uint32_t harness_last_value;
volatile uint32_t harness_last_status;
volatile uint32_t harness_last_restart_status;
volatile uint32_t harness_last_active_check;
volatile uint32_t harness_last_checkpoint_check;
volatile uint32_t harness_last_active_value;
volatile uint32_t harness_last_checkpoint_value;
volatile uint32_t harness_passes;
volatile uint32_t harness_failures;
volatile uint32_t harness_last_fault_target;
volatile checkpoint_record_t harness_c_checkpoint_state;

static uint32_t sample_initial_value(uint32_t iteration) {
    return 100u + ((iteration * 37u) % 700u);
}

static uint32_t sample_updated_value(uint32_t iteration) {
    return 100u + ((iteration * 53u + 211u) % 700u);
}

static checker_record_t sample_record(uint32_t value) {
    return checker_record_init(
        CHECKER_TAG_SAMPLE,
        value,
        0u,
        1000u,
        6u,
        16u);
}

__attribute__((used, noinline))
void harness_injection_point_after_mutation(void) {
    __asm__ volatile("nop");
}

__attribute__((used, noinline))
void harness_injection_point_after_commit(void) {
    __asm__ volatile("nop");
}

static void mirror_state(void) {
    const checkpoint_record_t *state =
        (const checkpoint_record_t *)&harness_c_checkpoint_state;
    harness_last_active_value = state->active.value;
    harness_last_checkpoint_value = state->checkpoint.value;
}

static void apply_pending_fault(void) {
    const uint32_t target = harness_fault_target;
    const uint32_t value = harness_fault_value;
    checkpoint_record_t *state = (checkpoint_record_t *)&harness_c_checkpoint_state;

    harness_last_fault_target = target;

    switch (target) {
    case HARNESS_FAULT_ACTIVE_VALUE:
        state->active.value = value;
        break;
    case HARNESS_FAULT_ACTIVE_LENGTH:
        state->active.length = value;
        break;
    case HARNESS_FAULT_ACTIVE_CHECKSUM:
        state->active.checksum ^= value;
        break;
    case HARNESS_FAULT_CHECKPOINT_VALUE:
        state->checkpoint.value = value;
        break;
    case HARNESS_FAULT_CHECKPOINT_CHECKSUM:
        state->checkpoint.checksum ^= value;
        break;
    case HARNESS_FAULT_ACTIVE_VALUE_AND_CHECKPOINT_CHECKSUM:
        state->active.value = value;
        state->checkpoint.checksum ^= 0x10u;
        break;
    default:
        break;
    }

    harness_fault_target = HARNESS_FAULT_NONE;
    mirror_state();
}

static void validate(uint32_t initial, uint32_t expected) {
    const uint32_t target = harness_last_fault_target;
    const uint32_t restart = harness_last_restart_status;
    const uint32_t active_check = harness_last_active_check;
    const uint32_t checkpoint_check = harness_last_checkpoint_check;
    const uint32_t active_value = harness_last_active_value;
    const uint32_t checkpoint_value = harness_last_checkpoint_value;

    switch (target) {
    case HARNESS_FAULT_ACTIVE_VALUE:
        if (restart == HARNESS_RESTART_RESTORED &&
            active_check == CHECKER_ERR_ABOVE_MAX &&
            checkpoint_check == CHECKER_OK &&
            active_value == initial &&
            checkpoint_value == initial) {
            harness_passes += 1;
        } else {
            harness_failures += 1;
        }
        return;
    case HARNESS_FAULT_ACTIVE_LENGTH:
        if (restart == HARNESS_RESTART_RESTORED &&
            active_check == CHECKER_ERR_INVALID_LENGTH &&
            checkpoint_check == CHECKER_OK &&
            active_value == initial &&
            checkpoint_value == initial) {
            harness_passes += 1;
        } else {
            harness_failures += 1;
        }
        return;
    case HARNESS_FAULT_ACTIVE_CHECKSUM:
        if (restart == HARNESS_RESTART_RESTORED &&
            active_check == CHECKER_ERR_INVALID_CHECKSUM &&
            checkpoint_check == CHECKER_OK &&
            active_value == initial &&
            checkpoint_value == initial) {
            harness_passes += 1;
        } else {
            harness_failures += 1;
        }
        return;
    case HARNESS_FAULT_ACTIVE_VALUE_AND_CHECKPOINT_CHECKSUM:
        if (restart == HARNESS_RESTART_RESTORE_FAILED &&
            active_check == CHECKER_ERR_ABOVE_MAX &&
            checkpoint_check == CHECKER_ERR_INVALID_CHECKSUM &&
            active_value == harness_fault_value) {
            harness_passes += 1;
        } else {
            harness_failures += 1;
        }
        return;
    default:
        if (restart == HARNESS_RESTART_COMMITTED &&
            active_check == CHECKER_OK &&
            checkpoint_check == CHECKER_OK &&
            active_value == expected &&
            checkpoint_value == expected) {
            harness_passes += 1;
        } else {
            harness_failures += 1;
        }
        return;
    }
}

void harness_main(void) {
    harness_stage = HARNESS_STAGE_BOOT;

    for (;;) {
        const uint32_t iteration = harness_iteration + 1u;
        const uint32_t initial = sample_initial_value(iteration);
        const uint32_t expected = sample_updated_value(iteration);
        const checker_record_t record = sample_record(initial);
        checkpoint_restart_result_t result;

        harness_iteration = iteration;
        harness_last_initial_value = initial;
        harness_last_expected = expected;
        harness_last_value = 0u;
        harness_last_status = HARNESS_STATUS_OK;
        harness_last_restart_status = HARNESS_RESTART_COMMITTED;
        harness_last_active_check = CHECKER_OK;
        harness_last_checkpoint_check = CHECKER_OK;
        harness_last_fault_target = HARNESS_FAULT_NONE;
        harness_c_checkpoint_state = checkpoint_record_init(record);

        harness_stage = HARNESS_STAGE_AFTER_CHECKPOINT;
        checkpoint_record_capture((checkpoint_record_t *)&harness_c_checkpoint_state);

        ((checkpoint_record_t *)&harness_c_checkpoint_state)->active.value = expected;
        checker_record_refresh_checksum(
            &((checkpoint_record_t *)&harness_c_checkpoint_state)->active);
        mirror_state();

        harness_stage = HARNESS_STAGE_AFTER_MUTATION;
        harness_injection_point_after_mutation();

        apply_pending_fault();

        harness_stage = HARNESS_STAGE_BEFORE_COMMIT;
        result = checkpoint_record_commit_or_restart(
            (checkpoint_record_t *)&harness_c_checkpoint_state);

        harness_last_restart_status = result.status;
        harness_last_active_check = result.active_check;
        harness_last_checkpoint_check = result.checkpoint_check;
        mirror_state();
        harness_last_value = harness_last_active_value;

        harness_stage = HARNESS_STAGE_AFTER_COMMIT;
        validate(initial, expected);
        harness_injection_point_after_commit();
    }
}

File: harness/e2e/zig/checkpoint_harness.zig
const checkpoint = @import("checkpoint");
const checker = @import("checker");
const abi = @import("abi");
const std = @import("std");

comptime {
    std.debug.assert(abi.restart.committed == @intFromEnum(checkpoint.RestartStatus.committed));
    std.debug.assert(abi.restart.restored == @intFromEnum(checkpoint.RestartStatus.restored));
    std.debug.assert(abi.restart.restore_failed == @intFromEnum(checkpoint.RestartStatus.restore_failed));
}

const check = struct {
    const ok: u32 = @intFromEnum(checker.CheckStatus.ok);
    const above_max: u32 = @intFromEnum(checker.CheckStatus.above_max);
    const invalid_length: u32 = @intFromEnum(checker.CheckStatus.invalid_length);
    const invalid_checksum: u32 = @intFromEnum(checker.CheckStatus.invalid_checksum);
};

export var harness_iteration: u32 = 0;
export var harness_stage: u32 = 0;
export var harness_fault_target: u32 = 0;
export var harness_fault_value: u32 = 0;
export var harness_last_expected: u32 = 0;
export var harness_last_initial_value: u32 = 0;
export var harness_last_value: u32 = 0;
export var harness_last_status: u32 = 0;
export var harness_last_restart_status: u32 = 0;
export var harness_last_active_check: u32 = 0;
export var harness_last_checkpoint_check: u32 = 0;
export var harness_last_active_value: u32 = 0;
export var harness_last_checkpoint_value: u32 = 0;
export var harness_passes: u32 = 0;
export var harness_failures: u32 = 0;
export var harness_last_fault_target: u32 = 0;

export var harness_zig_active_tag: u32 = 0;
export var harness_zig_active_value: u32 = 0;
export var harness_zig_active_min: u32 = 0;
export var harness_zig_active_max: u32 = 0;
export var harness_zig_active_length: u32 = 0;
export var harness_zig_active_capacity: u32 = 0;
export var harness_zig_active_checksum: u32 = 0;
export var harness_zig_checkpoint_tag: u32 = 0;
export var harness_zig_checkpoint_value: u32 = 0;
export var harness_zig_checkpoint_min: u32 = 0;
export var harness_zig_checkpoint_max: u32 = 0;
export var harness_zig_checkpoint_length: u32 = 0;
export var harness_zig_checkpoint_capacity: u32 = 0;
export var harness_zig_checkpoint_checksum: u32 = 0;

fn load(ptr: *const volatile u32) u32 {
    return ptr.*;
}

fn store(ptr: *volatile u32, value: u32) void {
    ptr.* = value;
}

fn sampleInitialValue(iteration: u32) u32 {
    return 100 + ((iteration *% 37) % 700);
}

fn sampleUpdatedValue(iteration: u32) u32 {
    return 100 + (((iteration *% 53) +% 211) % 700);
}

fn sampleRecord(value: u32) checker.CheckedRecord {
    return checker.CheckedRecord.init(
        .sample,
        value,
        0,
        1000,
        6,
        16,
    );
}

fn loadActiveRecord() checker.CheckedRecord {
    return .{
        .tag = load(&harness_zig_active_tag),
        .value = load(&harness_zig_active_value),
        .min = load(&harness_zig_active_min),
        .max = load(&harness_zig_active_max),
        .length = load(&harness_zig_active_length),
        .capacity = load(&harness_zig_active_capacity),
        .checksum = load(&harness_zig_active_checksum),
    };
}

fn loadCheckpointRecord() checker.CheckedRecord {
    return .{
        .tag = load(&harness_zig_checkpoint_tag),
        .value = load(&harness_zig_checkpoint_value),
        .min = load(&harness_zig_checkpoint_min),
        .max = load(&harness_zig_checkpoint_max),
        .length = load(&harness_zig_checkpoint_length),
        .capacity = load(&harness_zig_checkpoint_capacity),
        .checksum = load(&harness_zig_checkpoint_checksum),
    };
}

fn loadState() checkpoint.CheckpointedRecord {
    return .{
        .active = loadActiveRecord(),
        .checkpoint = loadCheckpointRecord(),
    };
}

fn mirrorState(state: *const checkpoint.CheckpointedRecord) void {
    store(&harness_last_active_value, state.active.value);
    store(&harness_last_checkpoint_value, state.checkpoint.value);
    store(&harness_zig_active_tag, state.active.tag);
    store(&harness_zig_active_value, state.active.value);
    store(&harness_zig_active_min, state.active.min);
    store(&harness_zig_active_max, state.active.max);
    store(&harness_zig_active_length, state.active.length);
    store(&harness_zig_active_capacity, state.active.capacity);
    store(&harness_zig_active_checksum, state.active.checksum);
    store(&harness_zig_checkpoint_tag, state.checkpoint.tag);
    store(&harness_zig_checkpoint_value, state.checkpoint.value);
    store(&harness_zig_checkpoint_min, state.checkpoint.min);
    store(&harness_zig_checkpoint_max, state.checkpoint.max);
    store(&harness_zig_checkpoint_length, state.checkpoint.length);
    store(&harness_zig_checkpoint_capacity, state.checkpoint.capacity);
    store(&harness_zig_checkpoint_checksum, state.checkpoint.checksum);
}

export fn harness_injection_point_after_mutation() callconv(.c) void {
    asm volatile ("nop // injection_point_after_mutation");
}

export fn harness_injection_point_after_commit() callconv(.c) void {
    asm volatile ("nop // injection_point_after_commit");
}

fn applyPendingFault(state: *checkpoint.CheckpointedRecord) void {
    const target = load(&harness_fault_target);
    const value = load(&harness_fault_value);

    store(&harness_last_fault_target, target);

    switch (target) {
        abi.fault.active_value => state.active.value = value,
        abi.fault.active_length => state.active.length = value,
        abi.fault.active_checksum => state.active.checksum ^= value,
        abi.fault.checkpoint_value => state.checkpoint.value = value,
        abi.fault.checkpoint_checksum => state.checkpoint.checksum ^= value,
        abi.fault.active_value_and_checkpoint_checksum => {
            state.active.value = value;
            state.checkpoint.checksum ^= 0x10;
        },
        else => {},
    }

    store(&harness_fault_target, abi.fault.none);
    mirrorState(state);
}

fn incrementPasses() void {
    store(&harness_passes, load(&harness_passes) +% 1);
}

fn incrementFailures() void {
    store(&harness_failures, load(&harness_failures) +% 1);
}

fn validate(initial: u32, expected: u32) void {
    const target = load(&harness_last_fault_target);
    const restart = load(&harness_last_restart_status);
    const active_check = load(&harness_last_active_check);
    const checkpoint_check = load(&harness_last_checkpoint_check);
    const active_value = load(&harness_last_active_value);
    const checkpoint_value = load(&harness_last_checkpoint_value);

    switch (target) {
        abi.fault.active_value => {
            if (restart == abi.restart.restored and
                active_check == check.above_max and
                checkpoint_check == check.ok and
                active_value == initial and
                checkpoint_value == initial)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        abi.fault.active_length => {
            if (restart == abi.restart.restored and
                active_check == check.invalid_length and
                checkpoint_check == check.ok and
                active_value == initial and
                checkpoint_value == initial)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        abi.fault.active_checksum => {
            if (restart == abi.restart.restored and
                active_check == check.invalid_checksum and
                checkpoint_check == check.ok and
                active_value == initial and
                checkpoint_value == initial)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        abi.fault.active_value_and_checkpoint_checksum => {
            if (restart == abi.restart.restore_failed and
                active_check == check.above_max and
                checkpoint_check == check.invalid_checksum and
                active_value == load(&harness_fault_value))
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
        else => {
            if (restart == abi.restart.committed and
                active_check == check.ok and
                checkpoint_check == check.ok and
                active_value == expected and
                checkpoint_value == expected)
            {
                incrementPasses();
            } else {
                incrementFailures();
            }
        },
    }
}

export fn harness_main() callconv(.c) noreturn {
    store(&harness_stage, abi.stage.boot);

    while (true) {
        const iteration = load(&harness_iteration) +% 1;
        const initial = sampleInitialValue(iteration);
        const expected = sampleUpdatedValue(iteration);
        var state = checkpoint.CheckpointedRecord.init(sampleRecord(initial));

        store(&harness_iteration, iteration);
        store(&harness_last_initial_value, initial);
        store(&harness_last_expected, expected);
        store(&harness_last_value, 0);
        store(&harness_last_status, abi.status.ok);
        store(&harness_last_restart_status, abi.restart.committed);
        store(&harness_last_active_check, check.ok);
        store(&harness_last_checkpoint_check, check.ok);
        store(&harness_last_fault_target, abi.fault.none);

        store(&harness_stage, abi.stage.after_checkpoint);
        _ = state.capture();

        state.active.value = expected;
        state.active.refreshChecksum();
        mirrorState(&state);

        store(&harness_stage, abi.stage.after_mutation);
        @call(.never_inline, harness_injection_point_after_mutation, .{});

        state = loadState();
        applyPendingFault(&state);

        store(&harness_stage, abi.stage.before_commit);
        const result = state.commitOrRestart();

        store(&harness_last_restart_status, result.status.code());
        store(&harness_last_active_check, result.active_check.code());
        store(&harness_last_checkpoint_check, result.checkpoint_check.code());
        mirrorState(&state);
        store(&harness_last_value, load(&harness_last_active_value));

        store(&harness_stage, abi.stage.after_commit);
        validate(initial, expected);
        @call(.never_inline, harness_injection_point_after_commit, .{});
    }
}

pub fn panic(msg: []const u8, trace: ?*std.builtin.StackTrace, ra: ?usize) noreturn {
    _ = msg;
    _ = trace;
    _ = ra;
    while (true) {}
}
